### PR TITLE
[ThirdParty] NCCL examples

### PR DESCRIPTION
**Pull request by feiwang3311**

### CODE REVIEW

File: doc/lms_code_walk.md [hunk 1]
```diff
@@ -23,8 +23,13 @@ into intermediate representations (IR). Instead, the IR is generated via executi
 typed expressions in the input program evaluate to LMS IR. This can be considered as the LMS frontend.
 Then the LMS backend compiles the LMS IR to target programs.
 
+## Overview
+
+The basic LMS IR is defined in this file:
 [LMS Core](main/scala/lms/core/backend.md)
 
+
+
```
Comment (feiwang3311) [inline doc/lms_code_walk.md:32]: I will fill this doc up :)

File: src/main/scala/lms/core/stub.scala [hunk 2]
```diff
@@ -1321,9 +1321,9 @@ abstract class DslSnippet[A:Manifest, B:Manifest] extends Dsl {
 }
 
 // Basic DslDriver for Scala CodeGen
-abstract class DslDriver[A:Manifest,B:Manifest] extends DslSnippet[A,B] with DslExp { q =>
+abstract class DslDriver[A:Manifest,B:Manifest] extends DslSnippet[A,B] with DslExp { self =>
   val codegen = new DslGen {
-    val IR: q.type = q
+    val IR: self.type = self
```
Comment (feiwang3311) [inline src/main/scala/lms/core/stub.scala:1326]: this is not strictly necessary. Just changing the name to make a point

**Comment by Kraks** [comment]
The `cu_examples` is a rather large code merge. A few minor comments on the organization: 1) why we need to commit `cu_examples` into lms-clean? 2) If so, would there be makefiles to compile those examples? 3) Those `.cu` files and `.h` files under `src/main/scala/lms/thirdparty/nccl/` seem to be better in somewhere else (eg `resources`), as `src/main/scala/...` is usually supposed to contain only `.scala` source code.

**Comment by feiwang3311** [comment]
> The `cu_examples` is a rather large code merge. A few minor comments on the organization: 1) why we need to commit `cu_examples` into lms-clean? 2) If so, would there be makefiles to compile those examples? 3) Those `.cu` files and `.h` files under `src/main/scala/lms/thirdparty/nccl/` seem to be better in somewhere else (eg `resources`), as `src/main/scala/...` is usually supposed to contain only `.scala` source code.

yeah. the cu_examples will be removed before merging.

File: src/main/scala/lms/thirdparty/mpi/mpi.scala [hunk 3]
```diff
@@ -25,17 +25,139 @@ trait MPIOps extends CMacro with LibStruct with LibFunction { b: Base =>
...
+      * datatype: data type of buffer (handle)
+      * root:     rank of broadcast root (integer)
+      * comm:     communicator (handle)
+   */
+  def mpi_bcast[T:Manifest](buffer: Rep[Array[T]], count: Rep[Int], datatype: Rep[MPIDataType], root: Rep[Int], comm: Rep[MPIComm]) =
+    libFunction[Int]("MPI_Bcast", Unwrap(buffer), Unwrap(count), Unwrap(datatype), Unwrap(root), Unwrap(comm))(Seq(0,2,4), Seq(0), Set[Int]())
+
+  // a variation of the function that uses the address as array
+  def mpi_bcast_one[T:Manifest](buffer: Rep[T], count: Rep[Int], datatype: Rep[MPIDataType], root: Rep[Int], comm: Rep[MPIComm]) =
+    libFunction[Int]("MPI_Bcast", Unwrap(buffer), Unwrap(count), Unwrap(datatype), Unwrap(root), Unwrap(comm))(Seq(0,2,4), Seq(0), Set[Int](0))
```
Comment (feiwang3311) [inline src/main/scala/lms/thirdparty/mpi/mpi.scala:125]: fix this

File: doc/lms_code_walk.md [hunk 4]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
 typed expressions in the input program evaluate to LMS IR. This can be considered as the LMS frontend.
 Then the LMS backend compiles the LMS IR to target programs.
 
+## Overview
+
+The basic components of LMS support includes:
+1. LMS IR: defining how the core LMS IR looks like.
+2. CodeGen: defining how to generate target language.
```
Comment (Kraks) [inline doc/lms_code_walk.md:30]:
```suggestion
2. CodeGen: defining how to generate code in the target language.
```

File: doc/lms_code_walk.md [hunk 5]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+3. FrontEnd: defining how to construct the LMS IR nicely.
+
+The basic LMS IR is defined in this file `lms-clean/src/main/scala/lms/core/backend.scala`:
 [LMS Core](main/scala/lms/core/backend.md)
 
+The simple typeless frontend is defined in this file `lms-clean/src/main/scala/lms/core/frontend.scala`:
 [LMS Simple Frontend](main/scala/lms/core/frontend.md)
 
+The iconic typed (`Rep[T]`) frondend is defined in this file `lms-clean/src/main/scala/lms/core/stub.scala`:
```
Comment (Kraks) [inline doc/lms_code_walk.md:39]:
```suggestion
The iconic typed (`Rep[T]`) frontend is defined in this file `lms-clean/src/main/scala/lms/core/stub.scala`:
```

File: doc/lms_code_walk.md [hunk 6]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
 [LMS Frontend](main/scala/lms/core/stub.md)
+
+The basic codegen support is in the file `lms-clean/src/main/scala/lms/core/codegen.scala`
+and the folder `lms-clean/src/main/scala/lms/core/codegen/*`.
+TODO(feiw): add a markdown file for codegen.
+
+
+### Extension
+
+Besides the basic LMS support mensioned above, LMS can be extended to support the following:
```
Comment (Kraks) [inline doc/lms_code_walk.md:49]:
```suggestion
Besides the basic LMS support mentioned above, LMS has included a number of extensions:
```

File: doc/lms_code_walk.md [hunk 7]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+The `trait CppCodeGen_List` is the trait for generating C++ code from the nodes
+created by `trait ListOps` and `trait ListOpsOpt`. Similarly, the most important
+function in this trait is `override def shallow` that generates code for each `Rep[List]` node.
+
+### ThirdParty Library
+
+Several examples of third party libraries are implemented here (`lms-clean/src/main/scala/lms/thirdparty`). The key functionality is provide by `lms-clean/src/main/scala/lms/thirdparty/c_lib_utils.scala`.
+
+In the `c_lib_utils.scala` file, we provide the following functionalities:
+1. Create a Macro by `def cmacro` function.
```
Comment (Kraks) [inline doc/lms_code_walk.md:81]:
```suggestion
1. Create a macro by the `cmacro` function.
```

File: doc/lms_code_walk.md [hunk 8]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+2. Create a C Struct by `def newStruct` function.
```
Comment (Kraks) [inline doc/lms_code_walk.md:82]:
```suggestion
2. Create a C struct by `newStruct` function.
```

File: doc/lms_code_walk.md [hunk 9]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+3. Create a call to library function by `def libFunction` function.
```
Comment (Kraks) [inline doc/lms_code_walk.md:83]:
```suggestion
3. Create a call to library functions by `libFunction` function.
```

File: doc/lms_code_walk.md [hunk 10]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+
+We only create three types of nodes for thirdparty libraries.
+1. The `def cmacro` function creates nodes with `op = "cmacro"`.
+2. The `def newStruct` function creates nodes with `op = "lib-struct"`.
+3. The `def libFunction` function creates nodes with `op = "lib-Function"`.
```
Comment (Kraks) [inline doc/lms_code_walk.md:88]:
```suggestion
1. The `cmacro` function creates nodes with `op = "cmacro"`.
2. The `newStruct` function creates nodes with `op = "lib-struct"`.
3. The `libFunction` function creates nodes with `op = "lib-Function"`.
```

File: doc/lms_code_walk.md [hunk 11]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+
+The codegen support those nodes are in this same file.
```
Comment (Kraks) [inline doc/lms_code_walk.md:90]:
```suggestion
The codegen supporting those nodes are in this same file.
```

File: doc/lms_code_walk.md [hunk 12]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+To mix in traits for the frontend support, just extend the `trait CLibs`.
+To mix in traits for the codegen support, just extend the `trait CCodeGenLibs`.
```
Comment (Kraks) [inline doc/lms_code_walk.md:92]: mix in -> mixin?

File: doc/lms_code_walk.md [hunk 13]
```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
...
+
+## Put Everything Together: Driver/Compiler
+
+With the support in frontends and codegen, we want to introduce a driver or a compiler that puts all the supports together and offers the programming environment for user application, code generation, compilation, and evaluation.
+
+The basic components of a driver or compiler include:
+1. All the needed frontend traits.
+2. All the needed codegen traits.
+3. An abstract `def snippet` that is the user application code
```
Comment (Kraks) [inline doc/lms_code_walk.md:101]:
```suggestion
3. An abstract method `snippet`, which will be overridden by the user application code
```

File: src/main/scala/lms/thirdparty/nccl/nccl.md [hunk 14]
```diff
@@ -0,0 +1,12 @@
+This folder tries to support NCCL (https://github.com/NVIDIA/nccl) as thirdparty library.
```
Comment (Kraks) [inline src/main/scala/lms/thirdparty/nccl/nccl.md:1]: maybe rename this file to `README.md` or move this to the doc folder?

File: src/main/scala/lms/thirdparty/nccl/nccl.scala [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+
+  import BaseTypeLess._
+  import PrimitiveTypeLess._
+  import ArrayTypeLess._
+  import SIZE_TTypeLess._
+  import CLibTypeLess._
+
+  class NCCL_RESULT
+
+  def NCCL_DATATYPE(m: Manifest[_])(implicit __pos: SourceContext) = m match {
```
Comment (Kraks) [inline src/main/scala/lms/thirdparty/nccl/nccl.scala:25]: could be refactored to 
```
if (ma == manifest[Int]) ...
else if (ma == manifest[Float]) ...
else throw ...
```

**Review by Kraks — APPROVED** [review]
Looks good to me, just a few minor comments/suggestions.

File: doc/lms_code_walk.md (hunk 12 above)
Comment (feiwang3311) [reply doc/lms_code_walk.md:92]: I guess `mix in` is verb and `mixin` is noun?

File: src/main/scala/lms/thirdparty/nccl/nccl.md (hunk 14 above)
Comment (feiwang3311) [reply src/main/scala/lms/thirdparty/nccl/nccl.md:1]: did

**Review by Kraks — APPROVED** [review]

**Pull request merged**